src/pow.rs

```diff
@@ -0,0 +1,1314 @@
+    /// the type.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    fn wrapping_mul(self, rhs: Self) -> Self {
+        let mut me = Uint256::default();
+        // TODO: be more efficient about this
+        for i in 0..8 {
+            let to_mul = (rhs >> (32 * i)).low_u32();
+            me = me.wrapping_add(self.mul_u32(to_mul) << (32 * i));
+        }
+        me
+    }
+
+    /// Panic-free bitwise shift-left; yields `self << mask(rhs)`, where `mask` removes any
+    /// high-order bits of `rhs` that would cause the shift to exceed the bitwidth of the type.
+    ///
+    /// Note that this is *not* the same as a rotate-left; the RHS of a wrapping shift-left is
+    /// restricted to the range of the type, rather than the bits shifted out of the LHS being
+    /// returned to the other end. We do not currently support `rotate_left`.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn wrapping_shl(self, rhs: usize) -> Self {
+        let shift = if rhs > 255 { rhs & 0x000000ff } else { rhs };
+        let shift = shift as usize;
+
+        let Uint256(ref original) = self;
+        let mut ret = [0u64; 4];
+        let word_shift = shift / 64;
+        let bit_shift = shift % 64;
+        for i in 0..4 {
+            // Shift
+            if bit_shift < 64 && i + word_shift < 4 {
+                ret[i + word_shift] += original[i] << bit_shift;
+            }
+            // Carry
+            if bit_shift > 0 && i + word_shift + 1 < 4 {
+                ret[i + word_shift + 1] += original[i] >> (64 - bit_shift);
+            }
+        }
+        Uint256(ret)
+    }
+
+    /// Panic-free bitwise shift-right; yields `self >> mask(rhs)`, where `mask` removes any
+    /// high-order bits of `rhs` that would cause the shift to exceed the bitwidth of the type.
+    ///
+    /// Note that this is *not* the same as a rotate-right; the RHS of a wrapping shift-right is
+    /// restricted to the range of the type, rather than the bits shifted out of the LHS being
+    /// returned to the other end. We do not currently support `rotate_right`.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    pub fn wrapping_shr(self, rhs: usize) -> Self {
+        let shift = if rhs > 255 { rhs & 0x000000ff } else { rhs };
+        let shift = shift as usize;
+
+        let Uint256(ref original) = self;
+        let mut ret = [0u64; 4];
+        let word_shift = shift / 64;
+        let bit_shift = shift % 64;
+        for i in word_shift..4 {
+            // Shift
+            ret[i - word_shift] += original[i] >> bit_shift;
+            // Carry
+            if bit_shift > 0 && i < 4 - 1 {
+                ret[i - word_shift] += original[i + 1] << (64 - bit_shift);
+            }
+        }
+        Uint256(ret)
+    }
+
+    #[cfg(test)]
+    fn bit(&self, index: usize) -> bool {
```

Comment:
Unimportant bikeshedding: `bit_at` seems better.

Reply:
Leaving open for input from others.

Reply:
I also kinda prefer `bit_at`

Reply:
Sweet, two for `bit_at`, that's good enough for me :)